### PR TITLE
feat : Register super admin user

### PR DESCRIPTION
**Pull request by marveldo**
## Description  : This update adds a new endpoint POST that registers super admin users
 ## How Has This Been Tested?

 1. **Unit Tests**:
    
* Implemented unit tests for the register super admin user endpoint
* I made sure it validates data received from the user properly
* I made it give the appropriate response on Success
 ##Screenshot of test
Screenshot 2024-07-24 214513
Screenshot 2024-07-25 125154
Screenshot 2024-07-25 125113
Screenshot 2024-07-25 121313

 ## Types of Changes
 
  * [ ]    Bug fix (non-breaking change which fixes an issue)
  * [x]    New feature (non-breaking change which adds functionality)
  * [ ]    Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

* [x]    My code follows the code style of this project.
* [x]    My change requires a change to the documentation.
* [x]    I have updated the documentation accordingly.
* [x]    I have read the CONTRIBUTING document.
* [x]    I have added tests to cover my changes.
* [x]    All new and existing tests passed.

 ## How to Run and Test the Project
### Prerequisites
 Ensure you have the following installed on your machine:

* Python 3.10
* PostgreSQL
* Git
* Virtual env (optional but recommended)

### Setup Instructions

1. **Clone the Repository**
 ```shell
 git clone <repository_url>
cd <repository_directory>
 ```
2. **Create and Activate a Virtual Environment (optional but recommended)**
```shell
python -m venv .venv
source .venv/bin/activate  # On Windows use `.venv\Scripts\activate`
```
3. **Install Dependencies**
```shell
pip install -r requirements.txt 
```
4. **Configure Environment Variables**
Create a `.env` file in the root directory and add the following environment variables:
 
PYTHON_ENV=dev
 
 
DB_TYPE=postgresql
DB_NAME=postgres
DB_USER=postgres
DB_PASSWORD=Winner174
DB_HOST="localhost"
DB_PORT=5432
MYSQL_DRIVER=
DB_URL=postgresql://postgres:Winner174@localhost:5432/postgres
SECRET_KEY = ""
ALGORITHM = HS256
ACCESS_TOKEN_EXPIRE_MINUTES = 10
JWT_REFRESH_EXPIRY=5
APP_URL=

MAIL_USERNAME=""
MAIL_PASSWORD=""
MAIL_FROM=""
MAIL_PORT=465
MAIL_SERVER="smtp.gmail.com"
 ```
5. **Run Database Migrations**
```shell
alembic upgrade head
 ```
 6. **Start the FastAPI Server**
 ```shell
 python main.py
```
   
The server will start at `http://127.0.0.1:7001`.

### Testing Instructions
 
1. **Run Unit Tests with Pytest**
```shell
pytest
``` 
This will run all the test cases and provide a report of the test results.

2. **Test the POST superadmin endpoint**
  
* Open Postman and create a new GET request to `http://127.0.0.1:7001/api/v1/superadmin/register`.
 * Send the request.
* You should receive a response indicating  that a user has been created and the details of the created user:
```json
  {
 {
  ```

 ```
         "status_code": 201,
           "success": true,
            "message": "User Created Successfully",
          "data": {
               "id": "066a175e-165d-7908-8000-4dcba2606466",
                "first_name": "Utibe",
                 "last_name": "Uboh",
                 "username": "Marvelous16",
                 "email": "utibesolomon17@gmail.com",
                     "created_at": "2024-07-24 22:45:05.093033+01:00"
```
 } }
 ```
        }
      ]
 }
 ``
 ### Notes
 
* Ensure your PostgreSQL database is running and accessible with the credentials provided in the `.env` file.
* If you encounter any issues, please check the logs for detailed error messages and verify that all dependencies are correctly installed.
```

### CODE REVIEW

**Review by joboy-dev — DISMISSED** [review]

File: tests/v1/test_admincreate.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    
+
+
+
+
+
+
+
+
+
```
Comment (Goketech) [inline tests/v1/test_admincreate.py:124]: too much space

**Review by Goketech — CHANGES_REQUESTED** [review]
review changes and format your pr message properly

File: tests/v1/test_admincreate.py (hunk 1 above)
Comment (marveldo) [reply tests/v1/test_admincreate.py]: ive fixed it @Goketech

**Review by Goketech — DISMISSED** [review]

**Review by joboy-dev — DISMISSED** [review]

**Review by joboy-dev — DISMISSED** [review]

File: api/v1/routes/superadmin.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+db_dependency = Annotated[Session , Depends(get_db)]
+
+
+@superadmin.post(path='/register', status_code=status.HTTP_201_CREATED)
+
+def register_admin(user : UserCreate , db : db_dependency):
+   
+    user_created = user_service.create(db=db, schema=user)
+    user_created.is_super_admin = True
+    db.commit()
```
Comment (Goketech) [inline api/v1/routes/superadmin.py:28]: did you test this, attach a post man screenshot of the working route.

what if a user has been created and we want to update role to user_admin,

create a function in the service folder that takes in the user schema, perform all necessary queries there

**Review by Goketech — CHANGES_REQUESTED** [review]

File: api/v1/routes/superadmin.py (hunk 2 above)
Comment (marveldo) [reply api/v1/routes/superadmin.py]: @Goketech  changes made boss i added  a new service that checks if that user exists and updats the user to as super user

**Review by joboy-dev — DISMISSED** [review]

**Review by Goketech — APPROVED** [review]

**Comment by Goketech** [comment]
@bolexs

**Review by joboy-dev — APPROVED** [review]

**Review by Goketech — DISMISSED** [review]

File: api/v1/schemas/user.py [hunk 3]
```diff
@@ -29,6 +29,9 @@ def password_validator(cls, value):
         if not re.match(r'^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$', value):
             raise ValueError('Password must be at least 8 characters long and contain at least one uppercase letter, one lowercase letter, one digit and one special character.')
         return value
+    
+
+
 
 # class SuccessResponseData(BaseModel):
```
Comment (bolexs) [inline api/v1/schemas/user.py:36]: why this?

Comment (bolexs) [reply api/v1/schemas/user.py:36]: please no comments can it be removed

Comment (marveldo) [reply api/v1/schemas/user.py]: @bolexs yes sir changes made

**Review by bolexs — APPROVED** [review]

**Pull request merged**